qir/qat/Rules/Factory.cpp

```diff
@@ -716,6 +718,82 @@ void RuleFactory::optimizeConstantResult()
     addRule({call("__quantum__rt__result_equal", "1"_cap = get_one, "2"_cap = get_one), replace_constant_result});
 }
 
+void RuleFactory::optimizeResultComparison()
+{
+    auto replace_comparison = [](Builder& builder, Value* val, Captures& cap, Replacements& replacements)
+    {
+        auto cond = llvm::dyn_cast<llvm::Instruction>(val);
+        if (cond == nullptr)
+        {
+            return false;
+        }
+        auto m1 = cap["m1"];
+        auto m2 = cap["m2"];
+
+        if (m1 == nullptr || m2 == nullptr)
+        {
+            return false;
+        }
+
+        auto orig_instr = llvm::dyn_cast<llvm::Instruction>(val);
+        if (orig_instr == nullptr)
+        {
+            return false;
+        }
+
+        auto module = orig_instr->getModule();
+        auto fnc    = module->getFunction("__quantum__qis__read_result__body");
+
+        if (!fnc)
+        {
+            std::vector<llvm::Type*> types;
+            types.resize(1);
+            types[0] = m1->getType();
+
+            auto return_type = llvm::Type::getInt1Ty(val->getContext());
+
+            llvm::FunctionType* fnc_type = llvm::FunctionType::get(return_type, types, false);
+            fnc                          = llvm::Function::Create(
+                                         fnc_type, llvm::Function::ExternalLinkage, "__quantum__qis__read_result__body", module);
+        }
+
+        builder.SetInsertPoint(llvm::dyn_cast<llvm::Instruction>(val));
+        auto c1       = builder.CreateCall(fnc, {m1});
+        auto c2       = builder.CreateCall(fnc, {m2});
+        auto new_cond = builder.CreateICmpEQ(c1, c2);
+
+        new_cond->takeName(cond);
+
+        for (auto& use : cond->uses())
+        {
+            llvm::User* user = use.getUser();
+            user->setOperand(use.getOperandNo(), new_cond);
+        }
+
+        cond->replaceAllUsesWith(new_cond);
+
+        // Deleting the previous condition and function to fetch one
+
+        replacements.push_back({cond, nullptr});
+
+        return true;
+    };
+
+    /*
+      Here is an example IR for which we want to make a match:
+
+        %result1 = call %Result* @__quantum__qis__m__body(%Qubit* %q0)
+        %result2 = call %Result* @__quantum__qis__m__body(%Qubit* %q1)
+        %0 = call i1 @__quantum__rt__result_equal(%Result* %result1, %Result* %result2)
+    */
+
+    // Variations of get_one
```

Comment:
Looks like this comment might be out of date, since these are not get_one variants.

Reply:
No this should not be there.